packit/local_project.py

```diff
@@ -562,3 +566,465 @@ def fetch(self, remote: str, refspec: Optional[str] = None, force: bool = False)
+        repo_name: Union[str, _CalculateType] = "",
+        # LocalProject "template"
+        local_project: Optional[LocalProject] = None,
+    ) -> LocalProject:
+        """Builds the LocalProject instance from the provided attributes.
+
+        Most arguments can either take an explicit value, None (or empty string) or
+        they can be passed the sentinel value CALCULATE to signal that this attribute
+        must be present in the final LocalProject object and therefore be computed
+        dynamically based on the rest of the information.
+
+        Args:
+            refresh: Whether attributes should be recalculated based on transitive relations.
+            remote: Name of the git remote to use.
+            pr_id: ID of the pull request to fetch and check out.
+            merge_pr: Whether the fetched pull request should be merged.
+            target_branch: Name of the branch the PR should be merged into.
+            git_repo: Instance of git.Repo.
+            working_dir: Path to the working directory of the project.
+            ref: Git ref, if set, it will be checked out.
+            git_project: ogr.GitProject (remote API)
+            git_service: ogr.Gitservice (tokens for remote API)
+            git_url: URL of the git project (used for cloning)
+            full_name: Full name of the remote git project (including namespace).
+            repo_name: Name of the remote git project.
+            namespace: Namespace of the remote git project.
+            local_project: LocalProject to use as the initial calculation state,
+                the other arguments will override information present in this LocalProject.
+
+        Returns:
+            The constructed LocalProject instance.
+        """
+        to_calculate: Set[str] = set()
+
+        def check_and_set(
+            calc_state: LocalProjectCalculationState, attr: str, value: Any
+        ):
+            if value is CALCULATE:
+                to_calculate.add(attr)
+            elif value is not None and value != "":
+                setattr(calc_state, attr, value)
+
+        if local_project:
+            state = LocalProjectCalculationState.from_local_project(local_project)
+        else:
+            state = LocalProjectCalculationState()
+        check_and_set(state, "remote", remote)
+        check_and_set(state, "git_repo", git_repo)
+        check_and_set(state, "working_dir", working_dir)
+        check_and_set(state, "ref", ref)
+        check_and_set(state, "git_project", git_project)
+        check_and_set(state, "git_service", git_service)
+        check_and_set(state, "git_url", git_url)
+        check_and_set(state, "full_name", full_name)
+        check_and_set(state, "namespace", namespace)
+        check_and_set(state, "repo_name", repo_name)
+        if refresh:
+            self._refresh_the_state(state, to_calculate)
+        # Do not refresh the built local project (the logic will go away)
```

Comment:
I am wondering if we can't get rid of this `refresh` argument... but yes, we might break something on the way.


(=Since we divide computing and state, we might not need to refresh anything. We should know if something needs to be calculated or not.)

Reply:
Thinking about that second time:
* I see you want to remove that from `LocalProject` once we migrate from direct instantiation. :+1:  (Once done, `LocalProject` might be converted to a data class...)
* But is there a use for having `refresh` in the new API? I think we like to be explicit so I don't see a use for that.

Reply:
Oh, now I see what you mean - good point :+1: